libminifi/src/utils/RegexUtils.cpp

```diff
@@ -0,0 +1,128 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "utils/RegexUtils.h"
+#include "Exception.h"
+#include <iostream>
+#include <vector>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+RegexUtils::RegexUtils(const std::string &value,
+                           const std::vector<RegexUtils::Mode> &mode)
+    : regexStr(value) {
+  // Create regex mode
+#if (__cplusplus > 201103L) || defined(_WIN32)
+  regex_mode = std::regex_constants::ECMAScript;
+#else
+  regex_mode = 0;
+#endif
+  for (const auto m : mode) {
+    switch (m) {
+      case Mode::EXTENDED:
+#if (__cplusplus > 201103L) || defined(_WIN32)
+        regex_mode |= std::regex_constants::extended;
+#else
+        regex_mode |= REG_EXTENDED;
+#endif
+        break;
+      case Mode::ICASE:
+#if (__cplusplus > 201103L) || defined(_WIN32)
+        regex_mode |= std::regex_constants::icase;
+#else
+        regex_mode |= REG_ICASE;
+#endif
+        break;
+    }
+  }
+  // Compile
+#if (__cplusplus > 201103L) || defined(_WIN32)
+  try {
+    compiledRegex = std::regex(regexStr, regex_mode);
+  } catch (const std::regex_error &e) {
+    throw e.what();
+  }
+#else
+  int err_code = regcomp(&compiledRegex, regexStr.c_str(), regex_mode);
+  if (err_code) {
+    char msg[100];
+    regerror(err_code, &compiledRegex, msg, sizeof(msg));
+    throw std::string(msg);
+  }
+  int maxGroups = std::count(regexStr.begin(), regexStr.end(), '(') + 1;
+  matches.resize(maxGroups);
+#endif
+}
+
+RegexUtils::~RegexUtils() {
+#if (__cplusplus > 201103L) || defined(_WIN32)
+
+#else
+  regfree(&compiledRegex);
+#endif
+}
+
+bool RegexUtils::canMatch(const std::string &pattern) {
+  results.clear();
+  if (pat == pattern) {
+    return false;
+  }
+  pat = pattern;
+#if (__cplusplus > 201103L) || defined(_WIN32)
+  if (std::regex_search(pattern, matches, compiledRegex)) {
+    for (const auto &m : matches) {
+      results.push_back(m.str());
+    }
+    suffix = matches.suffix();
+    return true;
+  }
+  return false;
+#else
+  if (regexec(&compiledRegex, pattern.c_str(), matches.size(),
+              matches.data(), 0) == 0) {
+    for (const auto &m : matches) {
+      if (m.rm_so == -1) {
+        break;
+      }
+      std::string s(pattern.begin() + m.rm_so, pattern.begin() + m.rm_eo);
+      results.push_back(s);
+    }
+    if ((size_t) matches[0].rm_eo >= pattern.size()) {
+      suffix = "";
+    } else {
+      suffix = pattern.substr(matches[0].rm_eo + 1);
+    }
+    return true;
+  }
+  return false;
+#endif
+}
+
+std::vector<std::string> RegexUtils::getResult() const { return results; }
```

Comment:
I think this can return with a const std::vector<std::string>& to prevent always copying the results.